src/sys/petsclib.cxx

```diff
@@ -47,5 +77,42 @@ void PetscLib::cleanup() {
   count = 0; // ensure that finalise is not called again later
 }
 
+void PetscLib::setPetscOptions(Options& options, std::string pass_options_prefix) {
+  // Pass all options in the section to PETSc
+  for (auto& i : options.getChildren()) {
+    if (not i.second.isValue()) {
+      throw BoutException("Found subsection %s in %s when reading PETSc options - only "
+          "values are allowed in the PETSc options, not subsections",
+          i.first.c_str(), options.str().c_str());
+    }
+    // Note, option names in the input file don't start with "-", but need to be passed
+    // to PETSc with "-" prepended
+    PetscErrorCode ierr;
+    if (lowercase(i.second) == "true") {
+      // PETSc flag with no value
+#if PETSC_VERSION_GE(3, 7, 0)
+      ierr = PetscOptionsSetValue(nullptr, ("-"+pass_options_prefix+i.first).c_str(),
```

Comment:
Is it worth pulling out the `"-"+pass_options_prefix+i.first).c_str()` part into a named variable to avoid repetition in the different branches? We could then also use this in the throw message below so we know which variable caused the issue.